### PR TITLE
add thermal control test plan

### PR DESCRIPTION
**Pull request by Junchao-Mellanox**

### CODE REVIEW

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+## Test Structure
+
+### Setup Configuration
+
+Since this feature is not related to traffic and network topology, default configuration on a T0 topology is good enough for this test.
+
+### Ansible and Pytest
+
+No new Ansible YML test case will be added. The test will reuse current [platform test](https://github.com/Azure/sonic-mgmt/tree/master/tests/platform) in sonic-mgmt. New pytest test cases will be added to [test_platform_info.py](https://github.com/Azure/sonic-mgmt/blob/master/tests/platform/test_platform_info.py). In addition, valid_policy.json and invalid_policy.json will be added as thermal policy configuration file for test purpose.
```
Comment (wangxin) [inline doc/pmon/sonic_thermal_control_test_plan.md]: YML -> YAML

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+## Scope
+
+This test is targeting a running SONIC system with fully functioning configuration. The purpose of the test is functional testing of thermal control on SONIC system, making sure that correct actions are executed once predefined conditions match.
+
+## Test Structure
+
+### Setup Configuration
+
+Since this feature is not related to traffic and network topology, default configuration on a T0 topology is good enough for this test.
```
Comment (wangxin) [inline doc/pmon/sonic_thermal_control_test_plan.md]: If there is no topology requirement, we can say that all the current topologies are supported. Please do not limit it to t0.

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+#### Procedure
+
+1. Testbed setup.
+2. Copy valid_policy.json to pmon docker and backup the original one.
+3. Restart pmon service to trigger thermal control daemon reload policy configuration file.
+4. Stop a PSU.
+5. Verify target speed of all Fans are set to value according to valid_policy.json.
+6. Resume the PSU.
+7. Verify target speed of all Fans are set to value according to valid_policy.json.
+
```
Comment (wangxin) [inline doc/pmon/sonic_thermal_control_test_plan.md:245]: Do we need to cover some variations in the policy? For example, different action, different algorithm?

**Review by wangxin — CHANGES_REQUESTED** [review]

File: doc/pmon/sonic_thermal_control_test_plan.md (hunk 3 above)
Comment (Junchao-Mellanox) [reply doc/pmon/sonic_thermal_control_test_plan.md:245]: Currently, we only support two actions: set FAN speed and control thermal algorithm. These two actions will be covered by current test. We could cover some variations in future if new actions added to our platform.

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+- FAN status monitor read FAN status via platform API every 60 seconds and save it to redis database. User can watch FAN status via command line `show platform fanstatus`.
+- Thermal status monitor read thermal status via platform API every 60 seconds and save it to redis database. User can watch thermal status via command line `show platform temperature`.
+- The thermal control policy is defined in a JSON file and loaded by thermal control daemon in pmon docker. Thermal control daemon collects thermal information and matches thermal conditions. Once some thermal conditions match, related thermal actions will be triggered. 
+
+More detail on thermal control feature can be found in this [document](https://github.com/keboliu/SONiC/blob/thermal_control_design/thermal-control-design.md).
+
+## Scope
+
+This test is targeting a running SONIC system with fully functioning configuration. The purpose of the test is functional testing of thermal control on SONIC system, making sure that FAN status and thermal status can shown to user and correct actions are executed once predefined thermal policy conditions match.
```
Comment (stephenxs) [inline doc/pmon/sonic_thermal_control_test_plan.md]: can be shown?

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+            "actions": [
+                {
+                    "type": "fan.all.set_speed",
+                    "speed": "-1"
+                }
+            ]
+        }
+    ]
+}
+```
```
Comment (stephenxs) [inline doc/pmon/sonic_thermal_control_test_plan.md:163]: should we add cases like invalid names, conditions and actions? 
but it's OK if they're covered by unit test.

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+#### Procedure
+
+1. Testbed setup
+2. Issue command `show platform temperature`.
+3. Record the command output
+4. Read thermal status via sysfs.
+5. Verify that command output matches thermal status from sysfs.
+
+> Note: Thermal status monitor update FAN status every 60 seconds, so the command output might not match FAN status from sysfs exactly. We may introduce some tolerance.
```
Comment (stephenxs) [inline doc/pmon/sonic_thermal_control_test_plan.md]: typo. should be thermal

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+### PSU Absence Test
+
+PSU absence test verifies that once any PSU absence, all FAN speed will be set to proper value according to policy file.
+
+#### Procedure
+
+1. Testbed setup.
+2. Copy valid_policy.json to pmon docker and backup the original one.
+3. Restart pmon service to trigger thermal control daemon reload policy configuration file.
+4. Stop two PSUs.
```
Comment (stephenxs) [inline doc/pmon/sonic_thermal_control_test_plan.md]: do you mean to power off the switch? 😀
regarding psu operations, I think you can refer the logic in PSU test cases which is something like:

    check whether all PSUs are turned on and exit if not(we don't want to turn off switch in this case)
    power off one PSU and check... 
    power on the PSU and power off the other and check... 
    finish

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+5. Wait for at least 60 seconds. Verify target speed of all FANs are set to 100% according to valid_policy.json.
```
Comment (stephenxs) [inline doc/pmon/sonic_thermal_control_test_plan.md]: to wait 60 seconds is risky. suggest waiting a bit longer to tolerance timing issues

**Review by stephenxs — CHANGES_REQUESTED** [review]
I'm not sure whether this pr contains fan and thermal monitoring test cases. if not I will looking forward to to. if yes, I suggest adding more test cases, like

1. simulate abnormal fan status like wrongspeed, absent, etc., and check whether expected log is displayed and led is set to the correct color
2. for thermal monitoring it's similar. 

log checking can be done by log analyzer and a normal fan status or thermal value can be simulated by unlinking the hwmgmt symbol links and putting normal file containing the abnormal values.

File: doc/pmon/sonic_thermal_control_test_plan.md (hunk 5 above)
Comment (Junchao-Mellanox) [reply doc/pmon/sonic_thermal_control_test_plan.md:163]: Yes, already covered by unit test.

File: doc/pmon/sonic_thermal_control_test_plan.md (hunk 7 above)
Comment (Junchao-Mellanox) [reply doc/pmon/sonic_thermal_control_test_plan.md]: Yes, I saw there is a psu_controller which can turn on/off the PSU.

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+  - [Show Thermal Status Test](#show-thermal-status-test)
+  - [FAN Test](#fan-test)
+  - [PSU Absence Test](#psu-absence-test)
+  - [Invalid Policy Format Load Test](#invalid-policy-format-load-test)
+  - [Invalid Policy Value Load Test](#invalid-policy-value-load-test)
+
+
+## Overview
+
+The purpose is to test functionality of thermal control feature on the SONIC switch DUT. The thermal control feature contains 3 major functions: FAN status monitor, thermal status monitor and thermal control policy management.
```
Comment (keboliu) [inline doc/pmon/sonic_thermal_control_test_plan.md]: SONIC --> SONiC

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+
+- FAN status monitor read FAN status via platform API every 60 seconds and save it to redis database. User can watch FAN status via command line `show platform fanstatus`.
```
Comment (keboliu) [inline doc/pmon/sonic_thermal_control_test_plan.md]: watch --> fetch?

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+- Thermal status monitor read thermal status via platform API every 60 seconds and save it to redis database. User can watch thermal status via command line `show platform temperature`.
```
Comment (keboliu) [inline doc/pmon/sonic_thermal_control_test_plan.md]: watch --> fetch?

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+- The thermal control policy is defined in a JSON file and loaded by thermal control daemon in pmon docker. Thermal control daemon collects thermal information and matches thermal conditions. Once some thermal conditions match, related thermal actions will be triggered. 
+
+More detail on thermal control feature can be found in this [document](https://github.com/keboliu/SONiC/blob/thermal_control_design/thermal-control-design.md).
+
+## Scope
+
+This test is targeting a running SONIC system with fully functioning configuration. The purpose of the test is functional testing of thermal control on SONIC system, making sure that FAN status and thermal status can be shown to user and correct actions are executed once predefined thermal policy conditions match.
```
Comment (keboliu) [inline doc/pmon/sonic_thermal_control_test_plan.md]: SONIC --> SONiC

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+Show thermal status test verifies that all thermal related information can be shown correctly via `show platform temperature`.
+
+#### Procedure
+
+1. Testbed setup.
+2. Unlink thermal related sysfs and fill fake data for "temperature", "high_threshold", "high_critical_threshold".
+3. Issue command `show platform temperature`.
+4. Record the command output.
+5. Verify that command output matches the fake data.
+6. Link thermal related sysfs.
```
Comment (keboliu) [inline doc/pmon/sonic_thermal_control_test_plan.md]: how to restore, re-link or config reload?
And we have a tool called "loganalyzer" which are widely used in test cases to check whether there is some error log or some specific log print out during the test, I suggest we can use it the analyze whether the warning msg printed out when the temperature value beyond the threshold.

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+
+### FAN Test
+
+FAN test verifies that proper action should be taken for conditions including: FAN absence, FAN over speed, FAN under speed.
+
+#### Procedure
+
+1. Testbed setup.
+2. Copy valid_policy.json to pmon docker and backup the original one.
+3. Restart pmon service to trigger thermal control daemon reload policy configuration file.
```
Comment (keboliu) [inline doc/pmon/sonic_thermal_control_test_plan.md]: how?

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+
+Show FAN status test verifies that all FAN related information can be shown correctly via `show platform fanstatus`.
+
+#### Procedure
+
+1. Testbed setup.
+2. Unlink FAN related sysfs and fill fake data for "presence", "speed", "status", "target_speed", "led status".
+3. Issue command `show platform fanstatus`.
+4. Record the command output.
+5. Verify that command output matches the fake data.
```
Comment (keboliu) [inline doc/pmon/sonic_thermal_control_test_plan.md]: fake --> mock?

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 16]
```diff
@@ -0,0 +1,250 @@
...
+
+Show thermal status test verifies that all thermal related information can be shown correctly via `show platform temperature`.
+
+#### Procedure
+
+1. Testbed setup.
+2. Unlink thermal related sysfs and fill fake data for "temperature", "high_threshold", "high_critical_threshold".
+3. Issue command `show platform temperature`.
+4. Record the command output.
+5. Verify that command output matches the fake data.
```
Comment (keboliu) [inline doc/pmon/sonic_thermal_control_test_plan.md]: fake --> mock?

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 17]
```diff
@@ -0,0 +1,250 @@
...
+
+Invalid policy value test verifies that thermal control daemon does not exit when loading a invalid_value_policy.json file. The thermal control daemon cannot perform any thermal policy in this case, but FAN monitor and thermal monitor should still work.
+
+#### Procedure
+
+1. Testbed setup.
+2. Copy invalid_value_policy.json to pmon docker and backup the original one.
+3. Restart pmon service to trigger thermal control daemon reload policy configuration file.
+4. Verify FAN status and thermal status still display normally. (By check command output of `show platform fanstatus` and `show platform temperature`)
+5. Recover the original policy configuration file and restart pmon service
```
Comment (keboliu) [inline doc/pmon/sonic_thermal_control_test_plan.md]: one thought about the restore,  maybe can combine the restore of each test case at the end of the test by a config reload, but this depends on how you implement the restore and whether the previous mock symbol link will impact the followed test.

**Review by keboliu — CHANGES_REQUESTED** [review]
as comments

File: doc/pmon/sonic_thermal_control_test_plan.md (hunk 6 above)
Comment (Junchao-Mellanox) [reply doc/pmon/sonic_thermal_control_test_plan.md]: removed the Note section

File: doc/pmon/sonic_thermal_control_test_plan.md (hunk 13 above)
Comment (Junchao-Mellanox) [reply doc/pmon/sonic_thermal_control_test_plan.md]: Re-link by "ln -s xxx" works, I tested it today. No need use "loganalyzer" because the "Show  Thermal Status Test" only mocks correct data, will add "loganalyzer" to negative test case.

File: doc/pmon/sonic_thermal_control_test_plan.md (hunk 14 above)
Comment (Junchao-Mellanox) [reply doc/pmon/sonic_thermal_control_test_plan.md]: By "systemctl restart pmon". Updated.

File: doc/pmon/sonic_thermal_control_test_plan.md (hunk 17 above)
Comment (Junchao-Mellanox) [reply doc/pmon/sonic_thermal_control_test_plan.md]: To unlink: just remove the symbol link and create a file with the same name.
To restore: remove the created file and add a soft link to hardware management file.
No need config reload.

Comment (stephenxs) [reply doc/pmon/sonic_thermal_control_test_plan.md]: if you are going to restore by re linking, it requires to restore the link ahead of removing it.

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 18]
```diff
@@ -0,0 +1,250 @@
...
+  - [PSU Absence Test](#psu-absence-test)
+  - [Invalid Policy Format Load Test](#invalid-policy-format-load-test)
+  - [Invalid Policy Value Load Test](#invalid-policy-value-load-test)
+
+
+## Overview
+
+The purpose is to test functionality of thermal control feature on the SONiC switch DUT. The thermal control feature contains 3 major functions: FAN status monitor, thermal status monitor and thermal control policy management.
+
+- FAN status monitor read FAN status via platform API every 60 seconds and save it to redis database. User can fetch FAN status via command line `show platform fanstatus`.
```
Comment (liat-grozovik) [inline doc/pmon/sonic_thermal_control_test_plan.md]: - suggest to put the function in ' ' for example: 'FAN status monitor'. same for the other functions.
- add s to read

Same comments for the other functions described below

Comment (liat-grozovik) [reply doc/pmon/sonic_thermal_control_test_plan.md]: the 60 seconds is based on the design, right? is it a configurable value? if not, have we considered it? if yes, should you read it and not use it as hard coded?

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 19]
```diff
@@ -0,0 +1,250 @@
...
+- Thermal status monitor read thermal status via platform API every 60 seconds and save it to redis database. User can fetch thermal status via command line `show platform temperature`.
+- The thermal control policy is defined in a JSON file and loaded by thermal control daemon in pmon docker. Thermal control daemon collects thermal information and matches thermal conditions. Once some thermal conditions match, related thermal actions will be triggered. 
+
+More detail on thermal control feature can be found in this [document](https://github.com/keboliu/SONiC/blob/thermal_control_design/thermal-control-design.md).
```
Comment (liat-grozovik) [inline doc/pmon/sonic_thermal_control_test_plan.md]: Suggest to rephrase:
A more detailed design and function description can be found it ...'

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 20]
```diff
@@ -0,0 +1,250 @@
...
+
+## Scope
+
+This test is targeting a running SONiC system with fully functioning configuration. The purpose of the test is functional testing of thermal control on SONiC system, making sure that FAN status and thermal status can be shown to user and correct actions are executed once predefined thermal policy conditions match.
+
```
Comment (liat-grozovik) [inline doc/pmon/sonic_thermal_control_test_plan.md:32]: Do you plan to verify different JSON files for different 'thermal policy'? if not, i suggest you will use the one defined in the system and some other functionality which can expose more flows to be validated.

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 21]
```diff
@@ -0,0 +1,250 @@
...
+## Test Structure
+
+### Setup Configuration
+
+Since this feature is not related to traffic and network topology, all current topology can be applied for this test.
+
+### Ansible and Pytest
+
+No new Ansible YAML test case will be added. The test will reuse current [platform test](https://github.com/Azure/SONiC-mgmt/tree/master/tests/platform) in SONiC-mgmt. New pytest test cases will be added to [test_platform_info.py](https://github.com/Azure/SONiC-mgmt/blob/master/tests/platform/test_platform_info.py). In addition, valid_policy.json,  invalid_format_policy.json and invalid_value_policy.json will be added as thermal policy configuration file for test purpose.
```
Comment (liat-grozovik) [inline doc/pmon/sonic_thermal_control_test_plan.md]: why referring to ansible? i believe  you should mention that it is based on platform test infra as additional cases.
BTW i think this document should be included in the platform test plan and not as a stand alone, we did update it few times with additional cases for the reboot cause.

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 22]
```diff
@@ -0,0 +1,250 @@
...
+
+#### Valid policy file
+
+In the valid policy file, two policies are defined. One is for "any PSU absence", the other is for "all FAN and PSU presence".
+
+- For "any PSU absence", all FAN speed need be set to 100% and thermal control algorithm need to be disabled.
```
Comment (liat-grozovik) [inline doc/pmon/sonic_thermal_control_test_plan.md]: Suggest to replace the following:
'For "any PSU absence", all FAN speed need be set to 100% and thermal control algorithm need to be disabled. 
For "all FAN and PSU presence", thermal control algorithm need to be enabled and FAN speed should be adjusted by it.'
with 
'In the case of  "any PSU absence", the expected behaviour based on the design and implementation is that FAN speed is set to 100% and thermal control algorithm is disabled.
In the case of "all FAN and PSU presence", the thermal control algorithm is enabled and the FAN speed is being adjusted by the thermal control.'

I am missing some more explanations of what and how to validate things such as
- what is fan speed 100%?
- what is expected fan speed when thermal control adjust it?

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 23]
```diff
@@ -0,0 +1,250 @@
...
+                }
+            ],
+            "actions": [
+                {
+                    "type": "thermal_control.control",
+                    "status": "false"
+                },
+                {
+                    "type": "fan.all.set_speed",
+                    "speed": "100"
```
Comment (liat-grozovik) [inline doc/pmon/sonic_thermal_control_test_plan.md:99]: is the platform API assumes % and not cycles?

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 24]
```diff
@@ -0,0 +1,250 @@
...
+                }
+            ],
+            "actions": [
+                {
+                    "type": "thermal_control.control",
+                    "status": "true"
+                },
+                {
+                    "type": "fan.all.set_speed",
+                    "speed": "60"
```
Comment (liat-grozovik) [inline doc/pmon/sonic_thermal_control_test_plan.md]: why 60? isnt the thermal control decide which will be the value based on the thermal zones?

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 25]
```diff
@@ -0,0 +1,250 @@
...
+```
+
+#### Invalid format policy file
+
+In invalid_format_policy.json, the file content is not valid JSON at all. A file with content "invalid" should be good for test purpose.
+
+#### Invalid value policy file
+
+In invalid_value_policy.json, the file content contains minus value of target speed. We couldn't cover all invalid value here because there are too many possibilities. The major purpose of this configuration file is to verify thermal control daemon would not crash while loading such an invalid configuration files. For other negative test cases, they are already covered by unit test. The invalid_value_policy.json content is like:
+
```
Comment (liat-grozovik) [inline doc/pmon/sonic_thermal_control_test_plan.md:135]: which unit testing are you referring to?
i think you should have a unit test like Stepan did for WD as he provided a general infrastructure and this can be also added to nightly regression

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 26]
```diff
@@ -0,0 +1,250 @@
...
+### Show FAN Status Test
+
+Show FAN status test verifies that all FAN related information can be shown correctly via `show platform fanstatus`.
+
+#### Procedure
+
+1. Testbed setup.
+2. Unlink FAN related sysfs and fill mock data for "presence", "speed", "status", "target_speed", "led status".
+3. Issue command `show platform fanstatus`.
```
Comment (liat-grozovik) [inline doc/pmon/sonic_thermal_control_test_plan.md:175]: isn't the show command reads from the REDIS DB? in this case how unlinking the FAS will be helpful? you need to probably to fill the DB with invalid info or missing info not just unlink it

Comment (liat-grozovik) [reply doc/pmon/sonic_thermal_control_test_plan.md:175]: what happen when thermal control is disabled? you should have such option and also need to cover enable/disable via CLI

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 27]
```diff
@@ -0,0 +1,250 @@
...
+
+### Show Thermal Status Test
+
+Show thermal status test verifies that all thermal related information can be shown correctly via `show platform temperature`.
+
+#### Procedure
+
+1. Testbed setup.
+2. Unlink thermal related sysfs and fill mock data for "temperature", "high_threshold", "high_critical_threshold".
+3. Issue command `show platform temperature`.
```
Comment (liat-grozovik) [inline doc/pmon/sonic_thermal_control_test_plan.md:188]: same comment as above regarding Redis DB

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 28]
```diff
@@ -0,0 +1,250 @@
...
+
+FAN test verifies that proper action should be taken for conditions including: FAN absence, FAN over speed, FAN under speed.
+
+#### Procedure
+
+1. Testbed setup.
+2. Copy valid_policy.json to pmon docker and backup the original one.
+3. Restart pmon service by command `systemctl restart pmon` to trigger thermal control daemon reload policy configuration file.
+4. Unlink FAN related sysfs and make mock data: first FAN absence.
+5. Wait for at least 65 seconds. Verify target speed of all FANs are set to 100% according to valid_policy.json. Verify there is a warning log for FAN absence.
```
Comment (liat-grozovik) [inline doc/pmon/sonic_thermal_control_test_plan.md:203]: where this 65 seconds come from? is there a config timeout we can read and dont use hard coded values?

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 29]
```diff
@@ -0,0 +1,250 @@
...
+6. Make mock data: first FAN presence.
+7. Wait for at least 65 seconds. Verify target speed of all FANs are set to 60% according to valid_policy.json. Verify there is a notice log for FAN presence.
```
Comment (liat-grozovik) [inline doc/pmon/sonic_thermal_control_test_plan.md]: i think 60% is not a good number.
when thermal control is disabled this is the default system value (at least in mellanox switches) thus if this comes from the config file, i suggest to use other value.

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 30]
```diff
@@ -0,0 +1,250 @@
...
+8. Make mock data: first FAN speed exceed threshold(speed < target speed), second FAN speed exceed threshold(speed > target speed).
```
Comment (liat-grozovik) [inline doc/pmon/sonic_thermal_control_test_plan.md:206]: how can you create moc data if thermal control is enabled and running in the system regardless of the moc?

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 31]
```diff
@@ -0,0 +1,250 @@
...
+### PSU Absence Test
+
+PSU absence test verifies that once any PSU absence, all FAN speed will be set to proper value according to policy file.
+
+#### Procedure
+
+1. Testbed setup.
+2. Copy valid_policy.json to pmon docker and backup the original one.
+3. Restart pmon service to trigger thermal control daemon reload policy configuration file.
+4. Turn off two PSUs.
```
Comment (liat-grozovik) [inline doc/pmon/sonic_thermal_control_test_plan.md]: this one you are using the PDU function, right? no simulation?

Comment (liat-grozovik) [reply doc/pmon/sonic_thermal_control_test_plan.md]: how can you turn off 2 PSUs? the system is not functional with no PSUs?
you probably need to shut 1, test, get it back, shut 2, test, get it back, test

File: doc/pmon/sonic_thermal_control_test_plan.md (hunk 3 above)
Comment (liat-grozovik) [reply doc/pmon/sonic_thermal_control_test_plan.md:245]: the limitation of the supported actions comes from hw mgmt or sonic implementation?

**Review by liat-grozovik — CHANGES_REQUESTED** [review]

File: doc/pmon/sonic_thermal_control_test_plan.md (hunk 3 above)
Comment (Junchao-Mellanox) [reply doc/pmon/sonic_thermal_control_test_plan.md:245]: It's by design. We design to support two actions for now.

File: doc/pmon/sonic_thermal_control_test_plan.md (hunk 17 above)
Comment (Junchao-Mellanox) [reply doc/pmon/sonic_thermal_control_test_plan.md]: Yes. I will save the original link in the test case.

File: doc/pmon/sonic_thermal_control_test_plan.md (hunk 18 above)
Comment (Junchao-Mellanox) [reply doc/pmon/sonic_thermal_control_test_plan.md]: It's based on design. User cannot configure it for now. Yes, we can do something to allow user configure it, either by policy json files or command line. What is your suggestion? I checked psud and xcvrd, they don't support configure poll interval for now.

File: doc/pmon/sonic_thermal_control_test_plan.md (hunk 22 above)
Comment (Junchao-Mellanox) [reply doc/pmon/sonic_thermal_control_test_plan.md]: We can verify FAN speed by read the sysfs file or by "show platform fanstatus". We don't know the expected fan speed when thermal control adjust it, but we can expect the sysfs of thermal control algorithm should be enabled.

File: doc/pmon/sonic_thermal_control_test_plan.md (hunk 28 above)
Comment (Junchao-Mellanox) [reply doc/pmon/sonic_thermal_control_test_plan.md:203]: Fan monitor, thermal monitor and thermal policy run every 60 seconds, so we wait longer than 60 which is 65. If we don't support user configure the running interval, we can't read it from any configuration.

File: doc/pmon/sonic_thermal_control_test_plan.md (hunk 30 above)
Comment (Junchao-Mellanox) [reply doc/pmon/sonic_thermal_control_test_plan.md:206]: Currently device status such as FAN speed, FAN presence status are read from sysfs by platform API. Those sysfs files are soft links to some "real files" controlled by hardware manager. So Stephen suggests to unlink those sysfs files so that device status will not be updated by hardware manager anymore, and we can mock data after that.

File: doc/pmon/sonic_thermal_control_test_plan.md (hunk 31 above)
Comment (Junchao-Mellanox) [reply doc/pmon/sonic_thermal_control_test_plan.md]: > this one you are using the PDU function, right? no simulation?

Yes, I plan to reuse the psu_controller in platform test package.

Comment (Junchao-Mellanox) [reply doc/pmon/sonic_thermal_control_test_plan.md]: > how can you turn off 2 PSUs? the system is not functional with no PSUs?
> you probably need to shut 1, test, get it back, shut 2, test, get it back, test

You are right. We can't shutdown two PSUs at the same time.

File: doc/pmon/sonic_thermal_control_test_plan.md (hunk 20 above)
Comment (Junchao-Mellanox) [reply doc/pmon/sonic_thermal_control_test_plan.md:32]: No, I am not going to verify different JSON files for different 'thermal policy'. The JSON files used by test is almost the same as the one defined in the system, except that it adds some actions to help verify the test result.

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 32]
```diff
@@ -0,0 +1,254 @@
...
+## Test Cases
+
+### Show FAN Status Test
+
+Show FAN status test verifies that all FAN related information can be shown correctly via `show platform fanstatus`.
+
+#### Procedure
+
+1. Testbed setup.
+2. Unlink FAN related sysfs and fill mock data for "presence", "speed", "status", "target_speed", "led status".
```
Comment (stepanblyschak) [inline doc/pmon/sonic_thermal_control_test_plan.md]: What do you mean by "unlink"? There is no standard sysfs path for such data in SONiC I think. Another platform could read it from registers directly and may not have sysfs entries.

File: doc/pmon/sonic_thermal_control_test_plan.md (hunk 30 above)
Comment (stepanblyschak) [reply doc/pmon/sonic_thermal_control_test_plan.md:206]: @Junchao-Mellanox The mocking procedure is very specific to mellanox platform. Not sure this will work for others.

File: doc/pmon/sonic_thermal_control_test_plan.md [hunk 33]
```diff
@@ -0,0 +1,254 @@
...
+> For switch who has only one PSU, step 6 and step 7 will be ignored.
+
+### Invalid Policy Format Load Test
+
+Invalid policy format test verifies that thermal control daemon does not exit when loading a invalid_format_policy.json file. The thermal control daemon cannot perform any thermal policy in this case, but FAN monitor and thermal monitor should still work.
+
+#### Procedure
+
+1. Testbed setup.
+2. Copy invalid_format_policy.json to pmon docker and backup the original one.
```
Comment (stepanblyschak) [inline doc/pmon/sonic_thermal_control_test_plan.md:241]: Add a log analyzer check that verifies there is an error?

**Review by stepanblyschak — CHANGES_REQUESTED** [review]
The described mocking seems to be very specific for hw-mgmt and mellanox platform.
Is this test only for mellanox platform?

File: doc/pmon/sonic_thermal_control_test_plan.md (hunk 32 above)
Comment (Junchao-Mellanox) [reply doc/pmon/sonic_thermal_control_test_plan.md]: Yes, "unlink" and most "mock" would only work for mellanox platform. As I can't mock data for other vendors, I will use code like " if dut.facts["asic_type"] in ["mellanox"]:" to make sure these test cases would not affect other vendor.

Comment (stepanblyschak) [reply doc/pmon/sonic_thermal_control_test_plan.md]: I would suggest not to use if mellanox conditions.
Instead, I suggest to define an interface to mock data from platform.
Each vendor will implement its own way to do that.
Mellanox will implement it by unlikning the sysfs file.
If implementation of  mocking is not available skip the test

Comment (Junchao-Mellanox) [reply doc/pmon/sonic_thermal_control_test_plan.md]: I will adjust my test code.

**Review by keboliu — APPROVED** [review]

**Review by stepanblyschak — APPROVED** [review]
